Invert ownership between PyFunction and THPFunction.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22983 Invert ownership between PyFunction and THPFunction.**

Fixes #1653, #14960 and #18619.

Whence this fix? Let's introduce the dramatis personae:

* THPFunction - a PyObject that represents a ctx object that is passed to our custom Python autograd functions. It contains a PyFunction
* PyFunction - a subclass of Function which says how to compute the actual backward function. Note that this is a closure in the sense that it is associated with a ctx object which actually contains the saved tensors. It contains a non-owning pointer to THPFunction

OK, do you see something wrong with this situation? The ownership here is totally backwards! It is PyFunction that should own the THPFunction, because a PyFunction is a closure that should own its context. To put it in other words, it is OK if a PyFunction gets freed before THPFunction, but not vice versa.

This sounds nice, but making this happen in practice is a bit trickier, because sometimes we use THPFunction to represent the PyFunction. In particular, we represent a `grad_fn` object in the Python binding level as the THPFunction (I can only imagine we did this because we wanted the object identity of `grad_fn` to line up with the class the user provided).  I did not have time to untangle this situation, so for this patch, in the cases where we're improperly treating a THPFunction as a PyFunction, we simply assume that the PyFunction is still live (it usually is), and yell at the user if it's not.

I constructed this patch in a few steps:
1. First, I flipped the ownership between PyFunction and THPFunction, but maintained a weak pointer from
THPFunction to PyFunction so that I could have an escape hatch when the existing code improperly treated THPFunction as an owning reference to PyFunction. In some cases your code might break because of this, but most of the time the PyFunction is live and everything works out (intuitively, this makes sense, since the ctx object should only really stay live as long as you're actually going to execute the backwards, which will keep the PyFunction live).
2. There are some cases where PyFunction is *provably* live at some point in time. In that case, don't bother extracting a strong reference from the weak pointer, just pass the strong pointer along directly!
3. For all the other cases where we can't prove it, write a long error message to help the user out when they run into that situation.

Some subtleties:
- PyFunction is a C++ object that refers to a PyObject.  This means it needs
  a custom deleter to handle deleting the PyObject, since you can't assume
  you have the GIL when it dies.
- The old test_gc_in_destructor failed our internal assert because we never
  actually ran a backwards, and thus never actually materialized PyFunction.
  I'm chalking this up as "misuse of API" and rewrote the test to not have
  this problem.
- Generalizing the issue above: this is BC-breaking for legacy autograd Functions
  in the following way: previously, the PyFunction for a THPFunction was generated
  immediately on construction of the Function `f` (which would eventually serve as
  the `grad_fn` of an object when you finally call `f(x)`.) This means that
  operations like `f.register_hook()` which previously worked prior to `f(x)` no longer
  work; you have to wait until you actually call the function to do registration.

I also removed dead Function::get_shared_ptr and Decref struct; these were previously needed for the crazy split refcounting scheme we had previously, but now are dead code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D16422209](https://our.internmc.facebook.com/intern/diff/D16422209)

ezyang committed:
Invert ownership between PyFunction and THPFunction.

Fixes #16532 and #14960.

This patch is a massive hack.  The way I constructed it was I flipped the
ownership between PyFunction and THPFunction, but maintained a weak pointer from
THPFunction to PyFunction so all existing code works.  Essentially, this patch
assumes that PyFunction stays live as long as you have a THPFunction: intuitively,
this makes sense, since the ctx object should only really stay live as long as
you're actually going to execute the backwards, which will keep the PyFunction
live.  But as you can see from the presently skipped tests (specifically,
test_hook_none), this is not always true.  But it seems to be true for the code
we care about, and that's enough for me!

Some subtleties:
- PyFunction is a C++ object that refers to a PyObject.  This means it needs
  a custom deleter to handle deleting the PyObject, since you can't assume
  you have the GIL when it dies.
- The old test_gc_in_destructor failed our internal assert because we never
  actually ran a backwards, and thus never actually materialized PyFunction.
  I'm chalking this up as "misuse of API" and rewrote the test to not have
  this problem.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
Subsumes #16888

ezyang commented:
There's some subsequent cleanup after this inversion that is possible, but I want to make sure this works first.

colesbury commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -238,12 +238,14 @@ auto PyFunction::get_shared_ptr() -> std::shared_ptr<Function> {
 {
-  for (const auto& hook : self->cdata.pre_hooks()) {
+  auto cdata = self->cdata.lock();
+  TORCH_INTERNAL_ASSERT(cdata);  // could do this optionally
```
How is this guaranteed?

ezyang replied:
It's not :) Here's what I have in my local copy:

```
 // Traverse and clear are required for supporting Python's GC cycle handling.
 static int THPFunction_traverse(THPFunction *self, visitproc visit, void *arg)
 {
-  for (const auto& hook : self->cdata.pre_hooks()) {
-    if (auto pyhook = dynamic_cast<PyFunctionPreHook*>(hook.get())) {
-      Py_VISIT(pyhook->dict);
+  auto cdata = self->cdata.lock();
+  // cdata could be null if someone constructed a legacy function but haven't
+  // actually called backward() on it yet.
+  if (cdata) {
+    for (const auto& hook : cdata->pre_hooks()) {
+      if (auto pyhook = dynamic_cast<PyFunctionPreHook*>(hook.get())) {
+        Py_VISIT(pyhook->dict);
+      }
     }
-  }
-  for (const auto& hook : self->cdata.post_hooks()) {
-    if (auto pyhook = dynamic_cast<PyFunctionPostHook*>(hook.get())) {
-      Py_VISIT(pyhook->dict);
+    for (const auto& hook : cdata->post_hooks()) {
+      if (auto pyhook = dynamic_cast<PyFunctionPostHook*>(hook.get())) {
+        Py_VISIT(pyhook->dict);
+      }
     }
   }
```

Waiting for CI to finish on the current iteration before I push though

colesbury replied:
Are you going to remove the other `TORCH_INTERNAL_ASSERT(cdata)` calls? I don't think any of them are safe.

mrshenli commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -635,7 +637,10 @@ PyObject *THPFunction_do_forward(THPFunction *self, PyObject *_inputs)
-  self->cdata.set_next_edges(std::move(input_info.next_edges));
+  Py_INCREF(self);
+  auto cdata = std::shared_ptr<PyFunction>(new PyFunction(THPObjectPtr((PyObject*)self)), deleteFunction);
+  self->cdata = cdata;
```
Is it possible that this method is called multiple times on the same `THPFunction`, so that in some invocations `self->cdata` is already non-NULL?

ezyang replied:
Yeah, in the legacy case it might be non-NULL. I actually have no idea what the "right" thing to do here is, because legacy autograd function behavior is nuts (all the better for @yf225 to delete them). Check this out:

```
macbook-pro-116:~ ezyang$ cat abb.py
import torch
from torch.autograd import Function
class Id(Function):
    def forward(self, x):
        self.save_for_backward(x)
        return x

    def backward(self, grad_x):
        x = self.saved_tensors
        print(x)
        return x

f = Id()
x1 = torch.zeros(1, requires_grad=True)
x2 = torch.ones(1, requires_grad=True)
print(f)
y = f(x1)
print(y.grad_fn)
z = f(x2)
print(z.grad_fn)
y.backward()

macbook-pro-116:~ ezyang$ python abb.py
<__main__.Id object at 0x116d58850>
<__main__.Id object at 0x116d58850>
<__main__.Id object at 0x116d58850>
(tensor([1.], requires_grad=True),)
```

The second call to f clobbers the saved tensors of the first call (in y), because f is *literally* the same `grad_fn` in both cases, even though this is totally wrong. I think the API is misused here, for a legacy Function, you should only call `__apply__` on it once.

On reflection, maybe the right thing is to not allocate a new `PyFunction` if there already is one. I can make this adjustment, but really this program is stupid and we shouldn't support it.

ezyang replied:
I've modified legacy to warn when this happens.

ezyang committed:
Update on "Invert ownership between PyFunction and THPFunction."

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22983 Invert ownership between PyFunction and THPFunction.**

Fixes #16532 and #14960.

This patch is a massive hack.  The way I constructed it was I flipped the
ownership between PyFunction and THPFunction, but maintained a weak pointer from
THPFunction to PyFunction so all existing code works.  Essentially, this patch
assumes that PyFunction stays live as long as you have a THPFunction: intuitively,
this makes sense, since the ctx object should only really stay live as long as
you're actually going to execute the backwards, which will keep the PyFunction
live.  But as you can see from the presently skipped tests (specifically,
test_hook_none), this is not always true.  But it seems to be true for the code
we care about, and that's enough for me!

Some subtleties:
- PyFunction is a C++ object that refers to a PyObject.  This means it needs
  a custom deleter to handle deleting the PyObject, since you can't assume
  you have the GIL when it dies.
- The old test_gc_in_destructor failed our internal assert because we never
  actually ran a backwards, and thus never actually materialized PyFunction.
  I'm chalking this up as "misuse of API" and rewrote the test to not have
  this problem.
- Generalizing the issue above: this is BC-breaking for legacy autograd Functions
  in the following way: previously, the PyFunction for a THPFunction was generated
  immediately on construction of the Function `f` (which would eventually serve as
  the `grad_fn` of an object when you finally call `f(x)`.) This means that
  operations like `f.register_hook()` which previously worked prior to `f(x)` no longer
  work; you have to wait until you actually call the function to do registration.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

mrshenli commented:
@ezyang I don't see any additional concerns with this PR (except the one @colesbury mentioned above), but I am not confident to give it a stamp either. Shall we bring more people to take a look?

apaszke commented on torch/csrc/autograd/python_function.h:
```diff
@@ -89,9 +97,11 @@ struct THPFunction {
+    // The actual PyFunction (in the autograd graph) that this data was
+    // saved for.  This field may be NULL (because a user can construct
+    // a THPFunction directly from Python), but when this field is non-NULL,
+    // it is guaranteed that cdata.lock()->obj == this
```
Why can't we enforce that Python `Function` objects always have their corresponding `PyFunction`s by allocating one whenever you instantiate an object? That way we could avoid having a weak pointer, and could hold `PyFunction*` will the invariant that it would always be non-null.

apaszke replied:
You also seem to be using the assumption that `.lock()` succeeds quite liberally in this patch, so this comment is highly confusing for me.

ezyang replied:
> Why can't we enforce that Python Function objects always have their corresponding PyFunctions by allocating one whenever you instantiate an object?

There are a few layers to this question, so I want to address them separately.

**In the non-legacy case, why can't we allocate a PyFunction in `THPFunction_new` so that `THPFunction` is always allocated with an associated `PyFunction`?** The proximal cause is dumb: it's because `THPFunction` only holds a weak pointer to `PyFunction`, so if you allocate a `PyFunction` inside `THPFunction_new`, the only pointer that would be keeping it live is a weak pointer, and it would immediately go dead once `THPFunction_new` returns. 

"But wait, Edward!" you might say, "Doesn't `THPFunction_apply` have this problem too?" No! We keep an owning reference to `PyFunction` for the entire body of this function, and eventually we will store an OWNING reference to `PyFunction` in the `grad_fn` of the Variables we return (in `process_outputs`); that owning reference will keep PyFunction live when we return. But there's no way to setup this owning reference in the constructor `THPFunction_new`, so we have to set it up somewhere else. It's a bit finely balanced, but in the non-legacy case, as long as you don't go constructing copies of `THPFunction` outside of `apply` (it is possible to do this, but not in the public API), we maintain the invariant that every THPFunction has a corresponding PyFunction.

**In the legacy case, why can't we guarantee that there is an associated PyFunction?** The first answer is, the legacy case is dumb and we shouldn't support it. But digging deeper, I think the root problem is that `THPFunction` (the `ctx` object) is what we expose to the world as `grad_fn`, when actually we should have exposed some sort of Python binding of `PyFunction`. So there is a confusion: we treat `THPFunction` as a Python binding as `PyFunction`, but we ALSO treat it as the `ctx` object we pass in when doing autograd. These two use cases are incompatible, but right now they're smooshed together in a single object.

The suggestion you've made here is treating `THPFunction` as just a Python binding for `PyFunction`. Which is reasonable, but a much more involved change.

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -1051,6 +1100,5 @@ std::shared_ptr<PyFunction> THPFunction_asFunction(THPFunction* self)
 
-  Py_INCREF((PyObject*)self);
-  return std::shared_ptr<PyFunction>(&self->cdata, Decref());
+  return self->cdata.lock();
```
How come we no longer need to do the incref/decref here? IIRC debug builds of Python report refcounting errors if we don't keep the C++/Python refcounts approximately in sync.

ezyang replied:
I don't know. I don't understand why it was previously needed. I guess I can build a debug build of Python and see if I still get refcounting errors. If I don't, would that be sufficient?

ezyang replied:
I have a debug build of PyTorch with this patch and it has not reported any refcounting errors.

apaszke replied:
Not a debug build of PyTorch, a debug build _of CPython_. I once made it such that we always use [a single `shared_ptr` to manage `PyFunction`](https://github.com/pytorch/pytorch/pull/1454), which has [started triggering Python's GC asserts](https://github.com/pytorch/pytorch/issues/1626) and the refcounting scheme [has been updated by @colesbury](https://github.com/pytorch/pytorch/pull/1705) accordingly. We should check that this patch does not suffer from the same issues as mine.

Also, there used to be this warning that you're not supposed to take weak pointers of `PyFunction`s because they can be managed by multiple shared_ptr blocks. That's no longer the case after this patch, right?

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -238,14 +238,19 @@ auto PyFunction::get_shared_ptr() -> std::shared_ptr<Function> {
+  auto cdata = self->cdata.lock();
+  // cdata could be null if someone constructed a legacy function but haven't
+  // actually called backward() on it yet.
+  if (cdata) {
```
nit: put the assignment into the `if` condition

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -222,7 +222,7 @@ auto PyFunction::name() const -> std::string {
   auto f = (THPFunction*) obj;
   auto name = std::string(Py_TYPE(f)->tp_name);
-  THPObjectPtr _legacy(PyObject_GetAttrString(obj, "_is_legacy"));
+  THPObjectPtr _legacy(PyObject_GetAttrString((PyObject*)obj, "_is_legacy"));  // CONST?!
```
use a nicer cast?

apaszke commented on test/test_nn.py:
```diff
@@ -4635,6 +4635,32 @@ def test_data_parallel_device_args(self):
 
+    @unittest.skipIf(not TEST_MULTIGPU, "multi-GPU not supported")
+    @skipIfRocm
+    def test_data_parallel_function_deletion(self):
```
Why is this related to this PR?

ezyang replied:
This tests the original bug that spawned this line of inquiry :)

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -635,7 +640,36 @@ PyObject *THPFunction_do_forward(THPFunction *self, PyObject *_inputs)
+      "from the second invocation will clobber the saved tensors from the "
+      "first invocation.  Please consider rewriting your autograd function "
+      "in the modern style; for information on the new format, please see: "
+      "https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd");
```
What if we use a legacy function once, free the backward, and try to apply it again? That wouldn't trigger this warning, so is that safe, or is it just another bug that is not covered by this PR?

ezyang replied:
Legacy is nuts so I don't really know for certain, but I think that it is technically safe. Because the backward is freed, there's nothing to clobber anymore.

ezyang replied:
We could tighten up the check by marking the legacy function as "consumed" once you apply it once (and error if you apply it again), but this is all best effort warning because we opted not to remove legacy functions this release cycle...

apaszke reviewed (commented):
This generally looks good and I think it makes sense (NB it would be good to link to the [explanation you wrote in another PR](https://github.com/pytorch/pytorch/pull/16888), because the description here is not very self-contained).

I am a bit worried about what would happen if someone was to hold on to the Python instance for too long (i.e. longer than the lifetime of the `PyFunction` from its use), and then try to access the attributes or reuse this. I think that would trigger your internal asserts, meaning they're not really that internal.

apaszke commented:
Also, I would be ok if we simply deleted the legacy paths today. They have been deprecated since at least 0.3 and it's just a pain to maintain them. Also, IIRC JIT will blow up if you try to mix them.

ezyang commented:
> I am a bit worried about what would happen if someone was to hold on to the Python instance for too long (i.e. longer than the lifetime of the PyFunction from its use), and then try to access the attributes or reuse this. I think that would trigger your internal asserts, meaning they're not really that internal.

Yeah, @colesbury complained about this in #22998 so I'm probably going to have to move around data little so it is guaranteed to stay live. None of the current test cases exercise the codepath but I certainly believe it's broken.

ezyang commented on test/test_autograd.py:
```diff
@@ -1674,12 +1675,38 @@ def test_gc_in_destructor(self):
+        with warnings.catch_warnings(record=True) as w:
+            z = f(x2)
+        self.assertIn('extending-torch-autograd', str(w[0].message))
+        y.backward()
```
`y.backward()` results in the clobber behavior. I'm not really interested in the functional correctness of clobbering (the behavior is nuts); I'm just mostly interested that we don't segfault in this situation.

ezyang commented:
> Also, I would be ok if we simply deleted the legacy paths today. They have been deprecated since at least 0.3 and it's just a pain to maintain them. Also, IIRC JIT will blow up if you try to mix them.

Yeah, me too XD. But IIUC they are somewhat widely used so we really should have an explicit deprecation notice.

ezyang committed:
Update on "Invert ownership between PyFunction and THPFunction."

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22983 Invert ownership between PyFunction and THPFunction.**

Fixes #16532 and #14960.

Let us introduce our dramatis personae:

* THPFunction - a PyObject that represents a ctx object that is passed to our custom Python autograd functions. It contains a PyFunction
* PyFunction - a subclass of Function which says how to compute the actual backward function. Note that this is a closure in the sense that it is associated with a ctx object which actually contains the saved tensors. It contains a non-owning pointer to THPFunction

OK, do you see something wrong with this situation? The ownership here is totally backwards! It is PyFunction that should own the THPFunction, because a PyFunction is a closure that should own its context. To put it in other words, it is OK if a PyFunction gets freed before THPFunction, but not vice versa.

This sounds nice, but making this happen in practice is a bit trickier, because there are cases where we do rely on ownership the wrong way.  This patch simply "assumes" that ownership goes the correct direction in practice.  The way I constructed it was I flipped the
ownership between PyFunction and THPFunction, but maintained a weak pointer from
THPFunction to PyFunction so all existing code works.  Essentially, this patch
assumes that PyFunction stays live as long as you have a THPFunction: intuitively,
this makes sense, since the ctx object should only really stay live as long as
you're actually going to execute the backwards, which will keep the PyFunction
live.  But as you can see from the presently skipped tests (specifically,
test_hook_none), this is not always true.  But it seems to be true for the code
we care about, and that's enough for me!

Some subtleties:
- PyFunction is a C++ object that refers to a PyObject.  This means it needs
  a custom deleter to handle deleting the PyObject, since you can't assume
  you have the GIL when it dies.
- The old test_gc_in_destructor failed our internal assert because we never
  actually ran a backwards, and thus never actually materialized PyFunction.
  I'm chalking this up as "misuse of API" and rewrote the test to not have
  this problem.
- Generalizing the issue above: this is BC-breaking for legacy autograd Functions
  in the following way: previously, the PyFunction for a THPFunction was generated
  immediately on construction of the Function `f` (which would eventually serve as
  the `grad_fn` of an object when you finally call `f(x)`.) This means that
  operations like `f.register_hook()` which previously worked prior to `f(x)` no longer
  work; you have to wait until you actually call the function to do registration.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Invert ownership between PyFunction and THPFunction."

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22983 Invert ownership between PyFunction and THPFunction.**

Fixes #16532 and #14960.

Whence this fix? Let's introduce the dramatis personae:

* THPFunction - a PyObject that represents a ctx object that is passed to our custom Python autograd functions. It contains a PyFunction
* PyFunction - a subclass of Function which says how to compute the actual backward function. Note that this is a closure in the sense that it is associated with a ctx object which actually contains the saved tensors. It contains a non-owning pointer to THPFunction

OK, do you see something wrong with this situation? The ownership here is totally backwards! It is PyFunction that should own the THPFunction, because a PyFunction is a closure that should own its context. To put it in other words, it is OK if a PyFunction gets freed before THPFunction, but not vice versa.

This sounds nice, but making this happen in practice is a bit trickier, because sometimes we use THPFunction to represent the PyFunction. In particular, we represent a `grad_fn` object in the Python binding level as the THPFunction (I can only imagine we did this because we wanted the object identity of `grad_fn` to line up with the class the user provided).  I did not have time to untangle this situation, so for this patch, in the cases where we're improperly treating a THPFunction as a PyFunction, we simply assume that the PyFunction is still live (it usually is), and yell at the user if it's not.

I constructed this patch in a few steps:
1. First, I flipped the ownership between PyFunction and THPFunction, but maintained a weak pointer from
THPFunction to PyFunction so that I could have an escape hatch when the existing code improperly treated THPFunction as an owning reference to PyFunction. In some cases your code might break because of this, but most of the time the PyFunction is live and everything works out (intuitively, this makes sense, since the ctx object should only really stay live as long as you're actually going to execute the backwards, which will keep the PyFunction live).
2. There are some cases where PyFunction is *provably* live at some point in time. In that case, don't bother extracting a strong reference from the weak pointer, just pass the strong pointer along directly!
3. For all the other cases where we can't prove it, write a long error message to help the user out when they run into that situation.

Some subtleties:
- PyFunction is a C++ object that refers to a PyObject.  This means it needs
  a custom deleter to handle deleting the PyObject, since you can't assume
  you have the GIL when it dies.
- The old test_gc_in_destructor failed our internal assert because we never
  actually ran a backwards, and thus never actually materialized PyFunction.
  I'm chalking this up as "misuse of API" and rewrote the test to not have
  this problem.
- Generalizing the issue above: this is BC-breaking for legacy autograd Functions
  in the following way: previously, the PyFunction for a THPFunction was generated
  immediately on construction of the Function `f` (which would eventually serve as
  the `grad_fn` of an object when you finally call `f(x)`.) This means that
  operations like `f.register_hook()` which previously worked prior to `f(x)` no longer
  work; you have to wait until you actually call the function to do registration.

I also removed dead Function::get_shared_ptr and Decref struct; these were previously needed for the crazy split refcounting scheme we had previously, but now are dead code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Invert ownership between PyFunction and THPFunction."

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22983 Invert ownership between PyFunction and THPFunction.**

Fixes #16532 and #14960.

Whence this fix? Let's introduce the dramatis personae:

* THPFunction - a PyObject that represents a ctx object that is passed to our custom Python autograd functions. It contains a PyFunction
* PyFunction - a subclass of Function which says how to compute the actual backward function. Note that this is a closure in the sense that it is associated with a ctx object which actually contains the saved tensors. It contains a non-owning pointer to THPFunction

OK, do you see something wrong with this situation? The ownership here is totally backwards! It is PyFunction that should own the THPFunction, because a PyFunction is a closure that should own its context. To put it in other words, it is OK if a PyFunction gets freed before THPFunction, but not vice versa.

This sounds nice, but making this happen in practice is a bit trickier, because sometimes we use THPFunction to represent the PyFunction. In particular, we represent a `grad_fn` object in the Python binding level as the THPFunction (I can only imagine we did this because we wanted the object identity of `grad_fn` to line up with the class the user provided).  I did not have time to untangle this situation, so for this patch, in the cases where we're improperly treating a THPFunction as a PyFunction, we simply assume that the PyFunction is still live (it usually is), and yell at the user if it's not.

I constructed this patch in a few steps:
1. First, I flipped the ownership between PyFunction and THPFunction, but maintained a weak pointer from
THPFunction to PyFunction so that I could have an escape hatch when the existing code improperly treated THPFunction as an owning reference to PyFunction. In some cases your code might break because of this, but most of the time the PyFunction is live and everything works out (intuitively, this makes sense, since the ctx object should only really stay live as long as you're actually going to execute the backwards, which will keep the PyFunction live).
2. There are some cases where PyFunction is *provably* live at some point in time. In that case, don't bother extracting a strong reference from the weak pointer, just pass the strong pointer along directly!
3. For all the other cases where we can't prove it, write a long error message to help the user out when they run into that situation.

Some subtleties:
- PyFunction is a C++ object that refers to a PyObject.  This means it needs
  a custom deleter to handle deleting the PyObject, since you can't assume
  you have the GIL when it dies.
- The old test_gc_in_destructor failed our internal assert because we never
  actually ran a backwards, and thus never actually materialized PyFunction.
  I'm chalking this up as "misuse of API" and rewrote the test to not have
  this problem.
- Generalizing the issue above: this is BC-breaking for legacy autograd Functions
  in the following way: previously, the PyFunction for a THPFunction was generated
  immediately on construction of the Function `f` (which would eventually serve as
  the `grad_fn` of an object when you finally call `f(x)`.) This means that
  operations like `f.register_hook()` which previously worked prior to `f(x)` no longer
  work; you have to wait until you actually call the function to do registration.

I also removed dead Function::get_shared_ptr and Decref struct; these were previously needed for the crazy split refcounting scheme we had previously, but now are dead code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Invert ownership between PyFunction and THPFunction."

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22983 Invert ownership between PyFunction and THPFunction.**

Fixes #16532 and #14960.

Whence this fix? Let's introduce the dramatis personae:

* THPFunction - a PyObject that represents a ctx object that is passed to our custom Python autograd functions. It contains a PyFunction
* PyFunction - a subclass of Function which says how to compute the actual backward function. Note that this is a closure in the sense that it is associated with a ctx object which actually contains the saved tensors. It contains a non-owning pointer to THPFunction

OK, do you see something wrong with this situation? The ownership here is totally backwards! It is PyFunction that should own the THPFunction, because a PyFunction is a closure that should own its context. To put it in other words, it is OK if a PyFunction gets freed before THPFunction, but not vice versa.

This sounds nice, but making this happen in practice is a bit trickier, because sometimes we use THPFunction to represent the PyFunction. In particular, we represent a `grad_fn` object in the Python binding level as the THPFunction (I can only imagine we did this because we wanted the object identity of `grad_fn` to line up with the class the user provided).  I did not have time to untangle this situation, so for this patch, in the cases where we're improperly treating a THPFunction as a PyFunction, we simply assume that the PyFunction is still live (it usually is), and yell at the user if it's not.

I constructed this patch in a few steps:
1. First, I flipped the ownership between PyFunction and THPFunction, but maintained a weak pointer from
THPFunction to PyFunction so that I could have an escape hatch when the existing code improperly treated THPFunction as an owning reference to PyFunction. In some cases your code might break because of this, but most of the time the PyFunction is live and everything works out (intuitively, this makes sense, since the ctx object should only really stay live as long as you're actually going to execute the backwards, which will keep the PyFunction live).
2. There are some cases where PyFunction is *provably* live at some point in time. In that case, don't bother extracting a strong reference from the weak pointer, just pass the strong pointer along directly!
3. For all the other cases where we can't prove it, write a long error message to help the user out when they run into that situation.

Some subtleties:
- PyFunction is a C++ object that refers to a PyObject.  This means it needs
  a custom deleter to handle deleting the PyObject, since you can't assume
  you have the GIL when it dies.
- The old test_gc_in_destructor failed our internal assert because we never
  actually ran a backwards, and thus never actually materialized PyFunction.
  I'm chalking this up as "misuse of API" and rewrote the test to not have
  this problem.
- Generalizing the issue above: this is BC-breaking for legacy autograd Functions
  in the following way: previously, the PyFunction for a THPFunction was generated
  immediately on construction of the Function `f` (which would eventually serve as
  the `grad_fn` of an object when you finally call `f(x)`.) This means that
  operations like `f.register_hook()` which previously worked prior to `f(x)` no longer
  work; you have to wait until you actually call the function to do registration.

I also removed dead Function::get_shared_ptr and Decref struct; these were previously needed for the crazy split refcounting scheme we had previously, but now are dead code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
I've resolved all of the code review comments.

ezyang committed:
Update on "Invert ownership between PyFunction and THPFunction."

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22983 Invert ownership between PyFunction and THPFunction.**

Fixes #1653, #14960 and #18619.

Whence this fix? Let's introduce the dramatis personae:

* THPFunction - a PyObject that represents a ctx object that is passed to our custom Python autograd functions. It contains a PyFunction
* PyFunction - a subclass of Function which says how to compute the actual backward function. Note that this is a closure in the sense that it is associated with a ctx object which actually contains the saved tensors. It contains a non-owning pointer to THPFunction

OK, do you see something wrong with this situation? The ownership here is totally backwards! It is PyFunction that should own the THPFunction, because a PyFunction is a closure that should own its context. To put it in other words, it is OK if a PyFunction gets freed before THPFunction, but not vice versa.

This sounds nice, but making this happen in practice is a bit trickier, because sometimes we use THPFunction to represent the PyFunction. In particular, we represent a `grad_fn` object in the Python binding level as the THPFunction (I can only imagine we did this because we wanted the object identity of `grad_fn` to line up with the class the user provided).  I did not have time to untangle this situation, so for this patch, in the cases where we're improperly treating a THPFunction as a PyFunction, we simply assume that the PyFunction is still live (it usually is), and yell at the user if it's not.

I constructed this patch in a few steps:
1. First, I flipped the ownership between PyFunction and THPFunction, but maintained a weak pointer from
THPFunction to PyFunction so that I could have an escape hatch when the existing code improperly treated THPFunction as an owning reference to PyFunction. In some cases your code might break because of this, but most of the time the PyFunction is live and everything works out (intuitively, this makes sense, since the ctx object should only really stay live as long as you're actually going to execute the backwards, which will keep the PyFunction live).
2. There are some cases where PyFunction is *provably* live at some point in time. In that case, don't bother extracting a strong reference from the weak pointer, just pass the strong pointer along directly!
3. For all the other cases where we can't prove it, write a long error message to help the user out when they run into that situation.

Some subtleties:
- PyFunction is a C++ object that refers to a PyObject.  This means it needs
  a custom deleter to handle deleting the PyObject, since you can't assume
  you have the GIL when it dies.
- The old test_gc_in_destructor failed our internal assert because we never
  actually ran a backwards, and thus never actually materialized PyFunction.
  I'm chalking this up as "misuse of API" and rewrote the test to not have
  this problem.
- Generalizing the issue above: this is BC-breaking for legacy autograd Functions
  in the following way: previously, the PyFunction for a THPFunction was generated
  immediately on construction of the Function `f` (which would eventually serve as
  the `grad_fn` of an object when you finally call `f(x)`.) This means that
  operations like `f.register_hook()` which previously worked prior to `f(x)` no longer
  work; you have to wait until you actually call the function to do registration.

I also removed dead Function::get_shared_ptr and Decref struct; these were previously needed for the crazy split refcounting scheme we had previously, but now are dead code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

colesbury reviewed (commented):
Looks good. Some minor comments below

ezyang commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -639,7 +640,33 @@ PyObject *THPFunction_do_forward(THPFunction *self, PyObject *_inputs)
   auto& input_info = info_pair.second;
   bool is_executable = input_info.is_executable;
-  self->cdata.set_next_edges(std::move(input_info.next_edges));
+  Py_INCREF(self);
```
It's so that I can give self as an owning reference to `THPObjectPtr`. Which means I'm leaking in the `TORCH_WARN` case. Better move it in!

ezyang commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -256,7 +260,11 @@ static int THPFunction_traverse(THPFunction *self, visitproc visit, void *arg)
+  // (otherwise the condition is trivially true).  Then there is a PyFunction
+  // which contains an owning reference to this object.  But we are only
+  // allowed to clear if all owning references are gone!  Contradiction.
+  TORCH_INTERNAL_ASSERT(!self->cdata.lock());
```
Thanks that is a very nice point :)

ezyang committed:
Update on "Invert ownership between PyFunction and THPFunction."

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22983 Invert ownership between PyFunction and THPFunction.**

Fixes #1653, #14960 and #18619.

Whence this fix? Let's introduce the dramatis personae:

* THPFunction - a PyObject that represents a ctx object that is passed to our custom Python autograd functions. It contains a PyFunction
* PyFunction - a subclass of Function which says how to compute the actual backward function. Note that this is a closure in the sense that it is associated with a ctx object which actually contains the saved tensors. It contains a non-owning pointer to THPFunction

OK, do you see something wrong with this situation? The ownership here is totally backwards! It is PyFunction that should own the THPFunction, because a PyFunction is a closure that should own its context. To put it in other words, it is OK if a PyFunction gets freed before THPFunction, but not vice versa.

This sounds nice, but making this happen in practice is a bit trickier, because sometimes we use THPFunction to represent the PyFunction. In particular, we represent a `grad_fn` object in the Python binding level as the THPFunction (I can only imagine we did this because we wanted the object identity of `grad_fn` to line up with the class the user provided).  I did not have time to untangle this situation, so for this patch, in the cases where we're improperly treating a THPFunction as a PyFunction, we simply assume that the PyFunction is still live (it usually is), and yell at the user if it's not.

I constructed this patch in a few steps:
1. First, I flipped the ownership between PyFunction and THPFunction, but maintained a weak pointer from
THPFunction to PyFunction so that I could have an escape hatch when the existing code improperly treated THPFunction as an owning reference to PyFunction. In some cases your code might break because of this, but most of the time the PyFunction is live and everything works out (intuitively, this makes sense, since the ctx object should only really stay live as long as you're actually going to execute the backwards, which will keep the PyFunction live).
2. There are some cases where PyFunction is *provably* live at some point in time. In that case, don't bother extracting a strong reference from the weak pointer, just pass the strong pointer along directly!
3. For all the other cases where we can't prove it, write a long error message to help the user out when they run into that situation.

Some subtleties:
- PyFunction is a C++ object that refers to a PyObject.  This means it needs
  a custom deleter to handle deleting the PyObject, since you can't assume
  you have the GIL when it dies.
- The old test_gc_in_destructor failed our internal assert because we never
  actually ran a backwards, and thus never actually materialized PyFunction.
  I'm chalking this up as "misuse of API" and rewrote the test to not have
  this problem.
- Generalizing the issue above: this is BC-breaking for legacy autograd Functions
  in the following way: previously, the PyFunction for a THPFunction was generated
  immediately on construction of the Function `f` (which would eventually serve as
  the `grad_fn` of an object when you finally call `f(x)`.) This means that
  operations like `f.register_hook()` which previously worked prior to `f(x)` no longer
  work; you have to wait until you actually call the function to do registration.

I also removed dead Function::get_shared_ptr and Decref struct; these were previously needed for the crazy split refcounting scheme we had previously, but now are dead code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -772,9 +812,21 @@ PyObject * THPFunction_do_backward(THPFunction *self, PyObject *args)
     }
-    THPUtils_assert(PyTuple_GET_SIZE(raw_grad_output) == self->cdata.num_inputs(),
+    auto cdata = self->cdata.lock();
+    // In obscure situations, cdata might be nullptr because it's expired.  THAT
```
Uhhhh are you sure? If it's expired then how did we get to `do_backward`? Isn't it the `PyFunction` that calls this method?

ezyang replied:
You can also just call `_do_backward` straight up. So the order of operations is, you save out the `grad_fn`, free the variables, and then call `grad_fn._do_backward`. It's dumb, not worth making it work better, but we do have an error check if this does happen :)

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -772,9 +812,21 @@ PyObject * THPFunction_do_backward(THPFunction *self, PyObject *args)
+      "was called.  This could occur if you manually called _do_backward on Function.  "
+      "In any case, this is very naughty!  If you absolutely need this to work, "
+      "try porting your code to use non-legacy autograd function, see: "
+      "https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd");
```
Oh lol, I think this is a bit paranoid, but w/e

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -818,17 +870,33 @@ PyObject * THPFunction_do_backward(THPFunction *self, PyObject *args)
-  self->cdata.add_pre_hook(std::move(hook));
+  auto cdata = self->cdata.lock();
+  TORCH_CHECK(cdata,
+    "Legacy autograd function had _register_hook_dict called before the function was "
```
This is a very bad error message to tell the user. Please figure out which methods call this one and write something that doesn't mention internal APIs. This gives completely no guidance about the potential fixes.

apaszke replied:
On the other hand it's legacy, so I guess I don't really care.

ezyang replied:
Well, it's easy enough to fix.

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -844,7 +912,12 @@ static PyObject *unpack_saved_variables(
+  // This is really a true assert, because we've already tested for the
+  // self->has_freed_buffers case at the beginning of this function:
+  // buffers are freed when PyFunction dies; if the buffers are not freed,
+  // PyFunction must be live.
```
Hmmm are you sure? I think those are orthogonal conditions. `has_freed_buffers` is only triggered by running backward, but you can have `cdata` get freed with no backward runs involved.

ezyang replied:
IIUC, it's a superset condition. If PyFunction is destructed, `has_freed_buffers` is guaranteed to be true, but not vice versa. That's good enough here.

apaszke reviewed (approved):
A few minor comments, but LGTM

ezyang commented on torch/csrc/autograd/functions/pybind.h:
```diff
@@ -11,21 +11,4 @@ namespace py = pybind11;
 namespace pybind11 { namespace detail {
 
-// handle Python <-> torch::autograd::Function conversions
-template <> struct type_caster<std::shared_ptr<torch::autograd::Function>> {
```
Well, Python side `grad_fn` poking still works after this patch, so that didn't exercise this code. I think this specifically only affects people who were writing pybind11 bindings for autograd functions. I have the tiniest violin.

ezyang commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -269,22 +285,14 @@ static int THPFunction_clear(THPFunction *self)
-  // trigger deletion of other objects, and they can reference this function,
-  // seeing it in a half-deleted state.
-  auto pre_hooks = std::move(self->cdata.pre_hooks());
-  auto post_hooks = std::move(self->cdata.post_hooks());
```
I think we're clear here. Proving that I don't need to do this is a big reason why we have the `TORCH_INTERNAL_ASSERT(self->cdata.expired());` assert at the top of the function. The claim is that PyFunction will always destruct first, and delete all of the hook objects, BEFORE we start deletion of THPFunction.

ezyang committed:
Update on "Invert ownership between PyFunction and THPFunction."

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22983 Invert ownership between PyFunction and THPFunction.**

Fixes #1653, #14960 and #18619.

Whence this fix? Let's introduce the dramatis personae:

* THPFunction - a PyObject that represents a ctx object that is passed to our custom Python autograd functions. It contains a PyFunction
* PyFunction - a subclass of Function which says how to compute the actual backward function. Note that this is a closure in the sense that it is associated with a ctx object which actually contains the saved tensors. It contains a non-owning pointer to THPFunction

OK, do you see something wrong with this situation? The ownership here is totally backwards! It is PyFunction that should own the THPFunction, because a PyFunction is a closure that should own its context. To put it in other words, it is OK if a PyFunction gets freed before THPFunction, but not vice versa.

This sounds nice, but making this happen in practice is a bit trickier, because sometimes we use THPFunction to represent the PyFunction. In particular, we represent a `grad_fn` object in the Python binding level as the THPFunction (I can only imagine we did this because we wanted the object identity of `grad_fn` to line up with the class the user provided).  I did not have time to untangle this situation, so for this patch, in the cases where we're improperly treating a THPFunction as a PyFunction, we simply assume that the PyFunction is still live (it usually is), and yell at the user if it's not.

I constructed this patch in a few steps:
1. First, I flipped the ownership between PyFunction and THPFunction, but maintained a weak pointer from
THPFunction to PyFunction so that I could have an escape hatch when the existing code improperly treated THPFunction as an owning reference to PyFunction. In some cases your code might break because of this, but most of the time the PyFunction is live and everything works out (intuitively, this makes sense, since the ctx object should only really stay live as long as you're actually going to execute the backwards, which will keep the PyFunction live).
2. There are some cases where PyFunction is *provably* live at some point in time. In that case, don't bother extracting a strong reference from the weak pointer, just pass the strong pointer along directly!
3. For all the other cases where we can't prove it, write a long error message to help the user out when they run into that situation.

Some subtleties:
- PyFunction is a C++ object that refers to a PyObject.  This means it needs
  a custom deleter to handle deleting the PyObject, since you can't assume
  you have the GIL when it dies.
- The old test_gc_in_destructor failed our internal assert because we never
  actually ran a backwards, and thus never actually materialized PyFunction.
  I'm chalking this up as "misuse of API" and rewrote the test to not have
  this problem.
- Generalizing the issue above: this is BC-breaking for legacy autograd Functions
  in the following way: previously, the PyFunction for a THPFunction was generated
  immediately on construction of the Function `f` (which would eventually serve as
  the `grad_fn` of an object when you finally call `f(x)`.) This means that
  operations like `f.register_hook()` which previously worked prior to `f(x)` no longer
  work; you have to wait until you actually call the function to do registration.

I also removed dead Function::get_shared_ptr and Decref struct; these were previously needed for the crazy split refcounting scheme we had previously, but now are dead code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>
